zjit/src/cruby_methods.rs

```diff
@@ -248,3 +249,13 @@ fn inline_array_aref(fun: &mut hir::Function, block: hir::BlockId, recv: hir::In
     }
     None
 }
+
+fn inline_string_getbyte(fun: &mut hir::Function, block: hir::BlockId, recv: hir::InsnId, args: &[hir::InsnId], state: hir::InsnId) -> Option<hir::InsnId> {
+    let &[index] = args else { return None; };
+    if fun.likely_a(index, types::Fixnum, state) {
+        let index = fun.coerce_to(block, index, types::Fixnum, state);
+        let result = fun.push_insn(block, hir::Insn::StringGetbyteFixnum { string: recv, index });
```

Comment:
I agree with using the Fixnum guard here, but it'd be nice to have a comment that explains why it's necessary (for backtraces, IIUC). It was not immediately obvious to me why you have to guard on the argument type, given that `rb_str_getbyte` is just the C method implementation that's also called for non-Fixnum cases.

Also, I'm not sure if this method deserves a dedicated HIR insn. Compared to things like `ArrayArefFixnum`, it's a fairly niche method that regular Ruby users would not use, unless you author something like a pure-Ruby protobuf implementation. Are you planning on adding constant folding on top of this for speeding up such benchmarks, or is it mainly for HIR presentation reasons, i.e. make it look better than `CCall` in HIR dumps?

Reply:
Added comments